src/libraries/System.Private.CoreLib/src/System/Int128.cs

```diff
@@ -1290,22 +1290,33 @@ public static Int128 Clamp(Int128 value, Int128 min, Int128 max)
         /// <inheritdoc cref="INumber{TSelf}.CopySign(TSelf, TSelf)" />
         public static Int128 CopySign(Int128 value, Int128 sign)
         {
-            Int128 absValue = value;
-
-            if (IsNegative(absValue))
-            {
-                absValue = -absValue;
-            }
-
-            if (IsPositive(sign))
+#if TARGET_32BIT
+            Int128 t = value;
+            if (Int128.IsPositive(sign))
             {
-                if (IsNegative(absValue))
+                if (Int128.IsNegative(t))
                 {
-                    Math.ThrowNegateTwosCompOverflow();
+                    t = -t;
+                    if (Int128.IsNegative(t))
+                    {
+                        Math.ThrowNegateTwosCompOverflow();
+                    }
                 }
-                return absValue;
+                return t;
             }
-            return -absValue;
+            if (Int128.IsPositive(t))
+            {
+                t = -t;
+            }
+            return t;
+#else
+            // Int128.MinValue == value && 0 <= sign
+            if ((long)value._upper == long.MinValue && value._lower == 0 && (long)sign._upper >= 0)
+            {
+                Math.ThrowNegateTwosCompOverflow();
+            }
+            return value * Math.SignOrOneIfZero((long)value._upper ^ (long)sign._upper);
```

Comment:
I'm still not a fan of this name nor of the cost for multiplication (which increases significantly as the size of the type does).

I think the entire logic here can be simplified down to:
```csharp
if (SameSign(value, sign))
{
    return value;
}
else if (value == Int128.MinValue)
{
    ThrowNegateTwosCompOverflow();
}

return -value;
```

where you have:
```csharp
bool SameSign(Int128 left, Int128 right)
{
    // All positive values have the most significant bit clear
    // All negative values have the most significant bit set
    //
    // (x ^ y) produces 0 if the bits are the same and 1 if they
    // differ. Therefore, (x ^ y) produces a positive value if the
    // signs are the same and a negative value if they differ.

    return (long)(left._upper ^ right._upper) >= 0;
}
```

This gives you up to the same 2 comparisons you currently have, but reduces the bit toggle to simply a two's complement operation (negation), which is far cheaper.

Reply:
The `SameSign` operation can even be optimized for `32-bit` platforms by only comparing the upper half of `_upper`, which is a much more reasonable platform specific optimization and keeps it heavily isolated away to only where needed.